### PR TITLE
Strided Tensor Support for Eager Mode

### PR DESCRIPTION
**Pull request by iK1D**
This PR constains:
- Add strides vector to onnxruntime::Tensor to support strided tensor
- Add MayStridedInput and MayStridedOutput to KernelDef to describe the strided tensor support status of kernel
- Allocation plan change according to whether kernel supporting strided tensor as input/output
- as_strided support in eager mode

### CODE REVIEW

File: include/onnxruntime/core/framework/tensor.h [hunk 1]
```diff
@@ -244,16 +250,39 @@ class Tensor final {
...
 
+  /**
+   * Get the strides of the tensor.
+   */
+  TensorShapeVector Strides() const;
+
+  /**
+   * Get pointer to stride array. Can be NULL if tensor is contiguous.
+   */
+  const int64_t* StridesPtr() const;
```
Comment (yuslepukhin) [inline include/onnxruntime/core/framework/tensor.h:261]: >const int64_t* StridesPtr() const; [](http://example.com/codeflow?start=2&length=34)

gsl:span<int64_t> ?

File: include/onnxruntime/core/framework/tensor.h [hunk 2]
```diff
@@ -244,16 +250,39 @@ class Tensor final {
...
+
+  /**
+   * Return if the tensor is contiguous.
+   */
+  bool IsContiguous() const { return is_contiguous_; }
```
Comment (yuslepukhin) [inline include/onnxruntime/core/framework/tensor.h:266]: >onst  [](http://example.com/codeflow?start=23&length=5)

noexcept

File: include/onnxruntime/core/framework/tensor.h [hunk 3]
```diff
@@ -244,16 +250,39 @@ class Tensor final {
   */
   size_t SizeInBytes() const;
 
+  /**
+   * Get the strides of the tensor.
+   */
+  TensorShapeVector Strides() const;
+
```
Comment (yuslepukhin) [inline include/onnxruntime/core/framework/tensor.h:257]: Can these two methods be the same returning gsl::span?

File: onnxruntime/core/framework/data_transfer.cc [hunk 4]
```diff
@@ -43,10 +47,20 @@ common::Status CPUDataTransfer::CopyTensor(const Tensor& src, Tensor& dst, int /
...
   }
-  // Copying only happens between two same size tensors.
-  ORT_ENFORCE(src.SizeInBytes() == dst.SizeInBytes());
-  memcpy(dst_data, src_data, src.SizeInBytes());
-  return Status::OK();
+
+  if (!src.IsContiguous() || !dst.IsContiguous()) {
+    auto dst_stride_vec = dst.Strides();
+    auto src_stride_vec = src.Strides();
+    onnxruntime::TensorShapeVector dst_stride{dst_stride_vec.begin(), dst_stride_vec.begin() + dst_stride_vec.size()};
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/data_transfer.cc:54]: >dst_stride_vec.begin() + dst_stride_vec.size() [](http://example.com/codeflow?start=70&length=46)

why not end()?

File: include/onnxruntime/core/framework/tensor.h [hunk 5]
```diff
@@ -39,8 +39,11 @@ class Tensor final {
   static std::unique_ptr<Tensor> Create(MLDataType p_type, const TensorShape& shape, std::shared_ptr<IAllocator> allocator) {
     return std::make_unique<Tensor>(p_type, shape, std::move(allocator));
   }
-  static std::unique_ptr<Tensor> Create(MLDataType p_type, const TensorShape& shape, void* p_data, const OrtMemoryInfo& alloc, ptrdiff_t offset = 0) {
-    return std::make_unique<Tensor>(p_type, shape, p_data, alloc, offset);
+
+  static std::unique_ptr<Tensor> Create(MLDataType p_type, const TensorShape& shape, void* p_data,
+                                        const OrtMemoryInfo& alloc, ptrdiff_t offset = 0,
+                                        const int64_t* p_strides = nullptr) {
```
Comment (yuslepukhin) [inline include/onnxruntime/core/framework/tensor.h:45]: >const int64_t* p_strides [](http://example.com/codeflow?start=40&length=24)

Can this be a gsl::span in both cases?

File: onnxruntime/core/framework/allocation_planner.cc [hunk 6]
```diff
@@ -357,6 +357,40 @@ class PlannerImpl {
         }
       }
     }
+
+    // If any output of the kernel can support strided tensor, and all its consumers' inputs also support
+    // strided tensors at the corresponding position, this output will generate a strided tensor
+    // and share the data from the corresponding input specified in MayStridedOutputsMap.
+    const std::vector<std::pair<int, int>>& may_strided_outputs_map = ci.kernel_def->MayStridedOutput();
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/allocation_planner.cc:364]: >std::vector<std::pair<int, int>> [](http://example.com/codeflow?start=10&length=32)

auto is better for future refactoring

File: onnxruntime/core/framework/allocation_planner.cc [hunk 7]
```diff
@@ -357,6 +357,40 @@ class PlannerImpl {
...
+    for (auto pair : may_strided_outputs_map) {
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/allocation_planner.cc:365]: >auto  [](http://example.com/codeflow?start=9&length=5)

ref

File: onnxruntime/core/framework/allocation_planner.cc [hunk 8]
```diff
@@ -357,6 +357,40 @@ class PlannerImpl {
...
+      if (pair.second == output_arg_num && pair.first >= 0 && static_cast<size_t>(pair.first) < input_args.size() &&
+          input_args[pair.first]->Exists()) {
+        bool can_strided = true;
+        for (auto it = node.OutputNodesBegin(); it != node.OutputNodesEnd(); ++it) {
+          const KernelCreateInfo& output_node_ci = GetKernelCreateInfo(kernel_create_info_map_, it->Index());
+          if (!output_node_ci.kernel_def) {
+            can_strided = false;
+            break;
+          }
+          const std::vector<int>& may_strided_inputs = output_node_ci.kernel_def->MayStridedInput();
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/allocation_planner.cc:375]: >std::vector<int> [](http://example.com/codeflow?start=16&length=16)

auto

File: onnxruntime/core/framework/allocation_planner.cc [hunk 9]
```diff
@@ -357,6 +357,40 @@ class PlannerImpl {
...
+          for (size_t i = 0; i < it->InputDefs().size(); ++i) {
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/allocation_planner.cc:376]: >or (size_t i = 0; i < it->InputDefs().size(); ++i) { [](http://example.com/codeflow?start=11&length=52)

What is the expected number of InputDefs() times strided_inputs?

File: orttraining/orttraining/eager/ort_util.h [hunk 10]
```diff
@@ -10,21 +10,29 @@
...
+void CreateMLValue(void* data_ptr,
+                   onnxruntime::MLDataType element_type,
+                   onnxruntime::TensorShape& shape,
+                   const OrtMemoryInfo& memory_info,
+                   OrtValue* p_mlvalue);
+
+void CreateMLValue(void* data_ptr,
+                   const OrtMemoryInfo& memory_info,
+                   onnxruntime::MLDataType element_type,
+                   const std::vector<int64_t>& dims,
```
Comment (yuslepukhin) [inline orttraining/orttraining/eager/ort_util.h:33]: >const std::vector<int64_t>& d [](http://example.com/codeflow?start=19&length=29)

For input use gsl::span<int64_t const>. This way you can pass any vector, or subspan there

**Comment by yuslepukhin** [comment]
> // Copyright (c) Microsoft Corporation. All rights reserved.

Why file names are not per Coding standard? We use .cc extensions

---
Refers to: orttraining/orttraining/eager/ort_util.cpp:1 in bb625fc. [](commit_id = bb625fcf35540ed56f33d81d2918a652259704b2, deletion_comment = False)

File: orttraining/orttraining/eager/ort_util.cpp [hunk 11]
```diff
@@ -49,6 +49,16 @@ void CreateMLValue(void* data_ptr,
...
 }
 
+void CreateMLValue(void* data_ptr, const OrtMemoryInfo& memory_info, onnxruntime::MLDataType element_type,
+                   const std::vector<int64_t>& dims, ptrdiff_t offset, const std::vector<int64_t>& strides,
+                   OrtValue* p_mlvalue) {
+  onnxruntime::TensorShape shape(dims);
+  std::unique_ptr<onnxruntime::Tensor> p_tensor = std::make_unique<onnxruntime::Tensor>(
+      element_type, shape, data_ptr, memory_info, offset, strides.empty() ? nullptr : &strides[0]);
+  p_mlvalue->Init(p_tensor.release(), onnxruntime::DataTypeImpl::GetType<onnxruntime::Tensor>(),
+                  onnxruntime::DataTypeImpl::GetType<onnxruntime::Tensor>()->GetDeleteFunc());
```
Comment (yuslepukhin) [inline orttraining/orttraining/eager/ort_util.cpp:59]: Would be be possible to use Tensor::InitOrtValue()?

File: orttraining/orttraining/eager/ort_aten.cpp [hunk 12]
```diff
@@ -378,6 +382,38 @@ at::Tensor empty_strided(
...
+  const at::Tensor& self, 
+  at::IntArrayRef size, 
+  at::IntArrayRef stride, 
+  c10::optional<int64_t> storage_offset) {
+  ORT_LOG_FN(self, size, stride, storage_offset);
+  auto& invoker = GetORTInvoker(self.device());
+  auto ort_input = create_ort_value(invoker, self);
+  auto* tensor = ort_input.GetMutable<onnxruntime::Tensor>();
+
+  std::vector<int64_t> stride_vec;
```
Comment (yuslepukhin) [inline orttraining/orttraining/eager/ort_aten.cpp:396]: >std::vector<int64_t> stride_vec [](http://example.com/codeflow?start=2&length=31)

TensorShapeVector

File: orttraining/orttraining/eager/ort_aten.cpp [hunk 13]
```diff
@@ -543,6 +579,67 @@ at::Tensor& add__Tensor(
...
+  }
+  if (l_end < l_start) {
+    l_end = l_start;
+  } else if (l_end >= shape[dim]) {
+    l_end = shape[dim];
+  }
+
+  auto byte_offset = ort_tensor->ByteOffset() + (l_start * strides[dim]) * ort_tensor->DataType()->Size();
+  auto len = l_end - l_start;
+  std::vector<int64_t> new_shape;
```
Comment (yuslepukhin) [inline orttraining/orttraining/eager/ort_aten.cpp:622]: >std::vector<int64_t> [](http://example.com/codeflow?start=2&length=20)

Do not use vector

File: orttraining/orttraining/eager/ort_util.cpp [hunk 14]
```diff
@@ -49,6 +49,16 @@ void CreateMLValue(void* data_ptr,
...
+}
+
 std::vector<int64_t> GetStrides(gsl::span<const int64_t> shape) {
```
Comment (yuslepukhin) [inline orttraining/orttraining/eager/ort_util.cpp:62]: >std::vector<int64_t> [](http://example.com/codeflow?start=0&length=20)

TesnorShapeVector

**Comment by pranavsharma** [comment]
Can you share a design doc for this change (as it touches the core inferencing areas as well)?

**Comment by iK1D** [comment]
> Can you share a design doc for this change (as it touches the core inferencing areas as well)?

There are two, one is at high level, other is in implementation level:
https://microsoft.sharepoint.com/:w:/t/ONNX2/EcF6n8cbHPdLosYFFwgFTj0BZzCS70rh66q-wo5H8Cq6Sg?e=Oas0iu
https://microsoft.sharepoint.com/:w:/t/ONNX2/ES8I3L0AF3BPoUggMmuJ_QMB5L-hujqzKywmEh5xIRwegA?e=qAybwL

**Comment by souptc** [comment]
> > // Copyright (c) Microsoft Corporation. All rights reserved.
> 
> Why file names are not per Coding standard? We use .cc extensions
> 
> Refers to: orttraining/orttraining/eager/ort_util.cpp:1 in [bb625fc](https://github.com/microsoft/onnxruntime/commit/bb625fcf35540ed56f33d81d2918a652259704b2). [](commit_id = [bb625fc](https://github.com/microsoft/onnxruntime/commit/bb625fcf35540ed56f33d81d2918a652259704b2), deletion_comment = False)

It is something history reason, the ort eager project originally is planned developed in a pytorch-extension repo, which following some pytorch code standard, like .cpp . But later we decide to move it into ort repo. I will start another PR to fix this.

**Review by souptc — APPROVED** [review]

**Pull request merged**